### PR TITLE
Nodejs week3

### PR DESCRIPTION
**Pull request by alina-kho**

### CODE REVIEW

File: src/backend/api/meals.js [hunk 1]
```diff
@@ -1,12 +1,161 @@
...
 router.get("/", async (request, response) => {
   try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
-    response.json(titles);
+    const maxPrice = request.query.maxPrice;
+    const availableReservations = request.query.availableReservations;
+    const title = request.query.title;
+    const createdAfter = request.query.createdAfter;
+    const limit = request.query.limit;
```
Comment (Divyajg) [inline src/backend/api/meals.js:15]: You can also do like `const { maxPrice, limit, ...} = request.query` then request query will be an object with all the queries.

File: src/backend/api/meals.js [hunk 2]
```diff
@@ -1,12 +1,161 @@
...
+    let results = knex("meals");
+
+    if (Object.keys(request.query).length == 0) {
+      const meals = await knex("meals").select("*");
```
Comment (Divyajg) [inline src/backend/api/meals.js]: `.select('*')` is suggested to use, instead you can specify the data you want.

Comment (Divyajg) [reply src/backend/api/meals.js]: use `const results` instead of `const meals `here.

File: src/backend/api/meals.js [hunk 3]
```diff
@@ -1,12 +1,161 @@
...
   try {
-    // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
-    response.json(titles);
+    const maxPrice = request.query.maxPrice;
+    const availableReservations = request.query.availableReservations;
+    const title = request.query.title;
+    const createdAfter = request.query.createdAfter;
+    const limit = request.query.limit;
+    let results = knex("meals");
```
Comment (Divyajg) [inline src/backend/api/meals.js]: if you use `const results` on line 19, you don't need this definition.

File: src/backend/api/meals.js [hunk 4]
```diff
@@ -1,12 +1,161 @@
...
+              knex.raw(
+                "meals.max_reservations - SUM(IFNULL(number_of_guests,0)) AS `reservations_available`"
+              )
+            )
+            .where("max_reservations", ">", "number_of_guests")
+            .groupBy("meals.id")
+            .having(
+              knex.raw(
+                "(meals.max_reservations - SUM(IFNULL(number_of_guests,0)))>0"
+              )
```
Comment (Divyajg) [inline src/backend/api/meals.js:87]: your solutions seems to be okay, I have seen many of your classmates had problem with this query. Try to discuss this in the class with the help of teacher/mentor.

File: src/backend/api/meals.js [hunk 5]
```diff
@@ -1,12 +1,161 @@
...
+          results = results
+            .leftJoin("reservations", "meals.id", "reservations.meal_id")
+            .select(
+              "meals.id",
+              "meals.title",
+              "meals.description",
+              "meals.price",
+              "meals.max_reservations",
+              knex.raw(
+                "SUM(IFNULL(number_of_guests,0)) AS `guests_registered`"
```
Comment (Divyajg) [inline src/backend/api/meals.js:76]: Try this `SUM(number_of_guests)`

File: src/backend/api/meals.js [hunk 6]
```diff
@@ -1,12 +1,161 @@
...
+        }
+      }
+
+      if (title) {
+        supportedParams = true;
+        results = results.where("title", "like", `%${title}%`).select("*");
+      }
+
+      if (createdAfter) {
+        if (new Date(createdAfter) != "Invalid Date") {
```
Comment (Divyajg) [inline src/backend/api/meals.js]: try `Date.parse()` instead of `new Date()`.

File: src/backend/api/meals.js [hunk 7]
```diff
@@ -1,12 +1,161 @@
...
+      title: request.body.title,
+      description: request.body.description,
+      location: request.body.location,
+      when: new Date(request.body.when),
+      max_reservations: request.body.maxReservations,
+      price: request.body.price,
+      created_date: new Date(),
+    };
+
+    const newMeal = await knex("meals").insert(meal);
```
Comment (Divyajg) [inline src/backend/api/meals.js:124]: `insert(request.body)` would be enough.

**Review by Divyajg — COMMENTED** [review]
Good job. 👍
I suggest you to minimise the code.

File: src/backend/api/meals.js (hunk 1 above)
Comment (alina-kho) [reply src/backend/api/meals.js:15]: Hi :) Sorry for the slow reply and thanks a lot for your feedback 🤓 That's a really nice point!

**Pull request merged**